### PR TITLE
XW-1370 add tests for document title

### PR DESCRIPTION
**Pull request by bognix**
@Wikia/x-wing

### CODE REVIEW

File: src/test/java/com/wikia/webdriver/common/core/url/UrlBuilder.java [hunk 1]
```diff
@@ -119,4 +119,12 @@ public static String getHostForWiki() {
   public static String getUrlForPage(String pageName) {
     return getHostForWiki() + pageName;
   }
+
+  public static String getUrl(String wikiName, String path) {
+    String environment = Configuration.getEnv();
+
+    environment = environment.equals("prod") ? "" : environment + ".";
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/common/core/url/UrlBuilder.java]: MAJOR Move the "prod" string literal on the left side of this string comparison. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS1132)

File: src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            }, {
+                new Page("starwars", "/wiki/Main_Page"), "Wookieepedia - Wikia"
+            }, {
+                new Page("starwars", "/wiki/Droid_starfighter"), "Droid starfighter - Wookieepedia - Wikia"
+            }, {
+                new Page("starwars", "/main/category/Future_films"), "Future_films"
+            }, {
+                new Page("starwars", "/main/section/Films"), "Films"
+            }, {
+                new Page("mediawiki119", "/d/f/203236/latest"), ""
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java]: MINOR Define a constant instead of duplicating this literal "mediawiki119" 3 times. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS1192)

File: src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            }, {
+                new Page("sktest123", "/wiki/User:Sktest"), "User:Sktest - Sktest123 Wiki - Wikia"
+            }, {
+                new Page("sktest123", "/wiki/File:THE_HOBBIT_Trailer_-_2012_Movie_-_Official_HD"), "Video - THE HOBBIT Trailer - 2012 Movie - Official HD - Sktest123 Wiki - Wikia"
+            }, {
+                new Page("sktest123", "/wiki/File:Giant_prominence_on_the_sun_erupted.jpg"), "Image - Giant prominence on the sun erupted.jpg - Sktest123 Wiki - Wikia"
+            }, {
+                new Page("sktest123", "/wiki/Special:NonExistingSpecialPage"), "Error - Sktest123 Wiki - Wikia"
+            }, {
+                new Page("es.pokemon", "/wiki/WikiDex"), "WikiDex - Wikia"
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java]: MINOR Define a constant instead of duplicating this literal "es.pokemon" 3 times. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS1192)

File: src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
+package com.wikia.webdriver.common.dataprovider;
+
+import com.wikia.webdriver.common.core.url.Page;
+import org.testng.annotations.DataProvider;
+
+public class MercurySEODataProvider {
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS1118)

File: src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            }, {
+                new Page("sktest123", "/wiki/Special:NonExistingSpecialPage"), "Error - Sktest123 Wiki - Wikia"
+            }, {
+                new Page("es.pokemon", "/wiki/WikiDex"), "WikiDex - Wikia"
+            }, {
+                new Page("es.pokemon", "/wiki/Lista_de_Pokémon"), "Lista de Pokémon - WikiDex - Wikia"
+            }, {
+                new Page("es.pokemon", "/wiki/Categoría:Regiones"), "Categoría:Regiones - WikiDex - Wikia"
+            }, {
+                new Page("starwars", "/wiki/Main_Page"), "Wookieepedia - Wikia"
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java]: MINOR Define a constant instead of duplicating this literal "starwars" 4 times. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS1192)

File: src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import com.wikia.webdriver.common.core.url.Page;
+import org.testng.annotations.DataProvider;
+
+public class MercurySEODataProvider {
+
+    @DataProvider
+    public static Object[][] htmlTitles() {
+        return new Object[][] {
+            {
+                new Page("sktest123", "/wiki/Sktest123_Wiki"), "Sktest123 Wiki - Wikia"
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java]: MINOR Define a constant instead of duplicating this literal "sktest123" 27 times. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS1192)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 6 issues:
* MAJOR 2 major
* MINOR 4 minor

Watch the comments in this conversation to review them.

File: src/test/java/com/wikia/webdriver/elements/mercury/components/Head.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  private WebElement documentTitle;
+
+  public Head(WebDriver driver) {
+    this.driver = driver;
+    PageFactory.initElements(driver, this);
+  }
+
+  public String getDocumentTitle() {
+    JavascriptExecutor js = (JavascriptExecutor) driver;
+    return (String) js.executeScript("return arguments[0].innerText", documentTitle);
```
Comment (tomasznapieralski) [inline src/test/java/com/wikia/webdriver/elements/mercury/components/Head.java]: you can use `.toString()` instead of casting

File: src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+import com.wikia.webdriver.common.core.helpers.Emulator;
+import com.wikia.webdriver.common.core.url.Page;
+import com.wikia.webdriver.common.dataprovider.MercurySEODataProvider;
+import com.wikia.webdriver.common.templates.TemplateNoFirstLoad;
+import com.wikia.webdriver.elements.common.Navigate;
+import com.wikia.webdriver.elements.mercury.components.Head;
+import org.testng.annotations.Test;
+
+@InBrowser(browser = Browser.CHROME, emulator = Emulator.GOOGLE_NEXUS_5)
+public class HTMLTitleTests extends TemplateNoFirstLoad {
```
Comment (tomasznapieralski) [inline src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java]: Consider using `NewTestTemplate`

File: src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    this.head = new Head(driver);
+    this.navigate = new Navigate(driver);
+  }
+
+  @Test(
+          groups = "htmlTitleSet",
+          dataProviderClass = MercurySEODataProvider.class,
+          dataProvider = "htmlTitles"
+  )
+  public void htmlTitleSet(Page page, String expectedTitle) {
```
Comment (tomasznapieralski) [inline src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java]: please rename it to have `mercury` prefix

File: src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    init();
+
+    String testPage = page.getUrl();
+    navigate.toUrl(testPage);
```
Comment (tomasznapieralski) [inline src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java]: merge those to methods please

File: src/test/java/com/wikia/webdriver/elements/mercury/components/Head.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
+package com.wikia.webdriver.elements.mercury.components;
+
+import org.openqa.selenium.JavascriptExecutor;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.PageFactory;
+
+public class Head {
```
Comment (ludwikkazmierczak) [inline src/test/java/com/wikia/webdriver/elements/mercury/components/Head.java]: Please extend WikiBasePageObject and get rid of `driver` parameter and `PageFactory` init  from constructor

File: src/test/java/com/wikia/webdriver/elements/mercury/components/Head.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+  @FindBy(css = "head title")
+  private WebElement documentTitle;
+
+  public Head(WebDriver driver) {
+    this.driver = driver;
+    PageFactory.initElements(driver, this);
+  }
+
+  public String getDocumentTitle() {
+    JavascriptExecutor js = (JavascriptExecutor) driver;
```
Comment (ludwikkazmierczak) [inline src/test/java/com/wikia/webdriver/elements/mercury/components/Head.java]: Once you extend WikiBasePageObject, casting to JavascriptExecutor is not needed

File: src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+import com.wikia.webdriver.elements.mercury.components.Head;
+import org.testng.annotations.Test;
+
+@InBrowser(browser = Browser.CHROME, emulator = Emulator.GOOGLE_NEXUS_5)
+public class HTMLTitleTests extends TemplateNoFirstLoad {
+
+  private Head head;
+  private Navigate navigate;
+
+  private void init() {
```
Comment (ludwikkazmierczak) [inline src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java]: Please use `@BeforeMethod(alwaysRun = true)` insted of `init()` inside of test body

File: src/test/java/com/wikia/webdriver/elements/mercury/components/Head.java (hunk 12 above)
Comment (ludwikkazmierczak) [inline src/test/java/com/wikia/webdriver/elements/mercury/components/Head.java]: Cant we just use: `driver.getTitle()` ?

File: src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java (hunk 8 above)
Comment (ludwikkazmierczak) [inline src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java]: With `TemplateNoFirstLoad` you wont be able to use AdsMock

File: src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java (hunk 9 above)
Comment (ludwikkazmierczak) [inline src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java]: I don't think prefixing is a good idea

File: src/test/java/com/wikia/webdriver/elements/mercury/components/Head.java (hunk 12 above)
Comment (bognix) [inline src/test/java/com/wikia/webdriver/elements/mercury/components/Head.java]: I haven't thought about that, however I would like to get the "raw" version of HTML title in case we have some errors - for instance trailing whitespaces.

I didn't extend WikiBasePageObject none of other components extends it. I had the impression that compoments should be as small and independent from other parts of code as possible. If you feel it's a problem I can change that however could you create guidelines when to create what? Current repo structure is very hard to follow and searching for up-to-date patterns isn't the easiest thing to do.

File: src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java (hunk 8 above)
Comment (bognix) [inline src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java]: Thanks - I'll rethink that.

File: src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java (hunk 9 above)
Comment (bognix) [inline src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java]: could you set up some guidelines how to name what? It's hard to follow. For now I'll leave it as it is and if you settle on something let me know - it's easy to fix.

File: src/test/java/com/wikia/webdriver/elements/mercury/components/Head.java (hunk 12 above)
Comment (ludwikkazmierczak) [inline src/test/java/com/wikia/webdriver/elements/mercury/components/Head.java]: WikiBasePageObject is the way to go for now, it's getting smaller and smaller every time, and we will clean that up finally. There (in `WBPO`) are few things we would like to keep - driver providing mechanisms, `PageFactory` init and basic `wait` and `jsActions` objects. Extending it will give you few bonuses:
1. You don't have to carry about drivers and object initialisation
2. You don't have to pass any drivers from a test body (components are aware which driver is the active one)
3. You have your wait and jsActions objects up front (two most common used ones)
All in all - less noise in components, and cleaner test (finally - no mythical `driver` object in test body :) )

 I know  we are missing updated guidelines, and that is a good point. We will get there ;)

File: src/test/java/com/wikia/webdriver/elements/mercury/components/Head.java (hunk 7 above)
Comment (ludwikkazmierczak) [inline src/test/java/com/wikia/webdriver/elements/mercury/components/Head.java]: It's better to cast in this situation, please keep it this way

File: src/test/java/com/wikia/webdriver/common/core/url/UrlBuilder.java (hunk 1 above)
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/common/core/url/UrlBuilder.java]: MAJOR Move the "prod" string literal on the left side of this string comparison. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS1132)

File: src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
+package com.wikia.webdriver.common.dataprovider;
+
+import com.wikia.webdriver.common.core.url.Page;
+import org.testng.annotations.DataProvider;
+
+public class MercurySEODataProvider {
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS1118)

File: src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+            }, {
+                new Page("starwars", "/wiki/Main_Page"), "Wookieepedia - Wikia"
+            }, {
+                new Page("starwars", "/wiki/Droid_starfighter"), "Droid starfighter - Wookieepedia - Wikia"
+            }, {
+                new Page("starwars", "/main/category/Future_films"), "Future_films"
+            }, {
+                new Page("starwars", "/main/section/Films"), "Films"
+            }, {
+                new Page("mediawiki119", "/d/f/203236/latest"), ""
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java]: MINOR Define a constant instead of duplicating this literal "mediawiki119" 3 times. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS1192)

File: src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+            }, {
+                new Page("sktest123", "/wiki/TestDisplayTitle"), "testing abc - Sktest123 Wiki - Wikia"
+            }, {
+                new Page("sktest123", "/wiki/Category:Premium_Videos"), "Category:Premium Videos - Sktest123 Wiki - Wikia"
+            }, {
+                new Page("sktest123", "/wiki/Category:Non-premium_Videos"), "Category:Non-premium Videos - Sktest123 Wiki - Wikia"
+            }, {
+                new Page("sktest123", "/wiki/Category:Premium"), "PremiumVideos - Sktest123 Wiki - Wikia"
+            }, {
+                new Page("es.pokemon", "/wiki/WikiDex"), "WikiDex - Wikia"
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java]: MINOR Define a constant instead of duplicating this literal "es.pokemon" 3 times. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS1192)

File: src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+import com.wikia.webdriver.common.core.url.Page;
+import org.testng.annotations.DataProvider;
+
+public class MercurySEODataProvider {
+
+    @DataProvider
+    public static Object[][] htmlTitles() {
+        return new Object[][] {
+            {
+                new Page("sktest123", "/wiki/Sktest123_Wiki"), "Sktest123 Wiki - Wikia"
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java]: MINOR Define a constant instead of duplicating this literal "sktest123" 6 times. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS1192)

File: src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+            }, {
+                new Page("sktest123", "/wiki/Category:Premium"), "PremiumVideos - Sktest123 Wiki - Wikia"
+            }, {
+                new Page("es.pokemon", "/wiki/WikiDex"), "WikiDex - Wikia"
+            }, {
+                new Page("es.pokemon", "/wiki/Lista_de_Pokémon"), "Lista de Pokémon - WikiDex - Wikia"
+            }, {
+                new Page("es.pokemon", "/wiki/Categoría:Regiones"), "Categoría:Regiones - WikiDex - Wikia"
+            }, {
+                new Page("starwars", "/wiki/Main_Page"), "Wookieepedia - Wikia"
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/common/dataprovider/MercurySEODataProvider.java]: MINOR Define a constant instead of duplicating this literal "starwars" 4 times. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS1192)

File: src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+import com.wikia.webdriver.common.core.annotations.InBrowser;
+import com.wikia.webdriver.common.core.drivers.Browser;
+import com.wikia.webdriver.common.core.helpers.Emulator;
+import com.wikia.webdriver.common.core.url.Page;
+import com.wikia.webdriver.common.dataprovider.MercurySEODataProvider;
+import com.wikia.webdriver.common.templates.TemplateNoFirstLoad;
+import com.wikia.webdriver.elements.common.Navigate;
+import com.wikia.webdriver.elements.mercury.components.Head;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.BeforeTest;
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java]: MINOR Remove this unused import 'org.testng.annotations.BeforeTest'. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 10 issues:
* MAJOR 2 major
* MINOR 8 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Remove this unused import 'java.util.Date'.](https://github.com/Wikia/selenium-tests/blob/ef93c67f9de79c18de4fb7fe46d7f8b52b1fa3cd/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/wam/WamPageObject.java#L21) [rule](http://sonar-s1/coding_rules#rule_key=squid%3AUselessImportCheck)
* MINOR [Remove this unused import 'org.openqa.selenium.JavascriptExecutor'.](https://github.com/Wikia/selenium-tests/blob/ef93c67f9de79c18de4fb7fe46d7f8b52b1fa3cd/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/wam/WamPageObject.java#L12) [rule](http://sonar-s1/coding_rules#rule_key=squid%3AUselessImportCheck)
* MINOR [Remove this unused import 'java.text.SimpleDateFormat'.](https://github.com/Wikia/selenium-tests/blob/ef93c67f9de79c18de4fb7fe46d7f8b52b1fa3cd/src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/wam/WamPageObject.java#L19) [rule](http://sonar-s1/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+    this.head = new Head();
+    this.navigate = new Navigate(driver);
+  }
+
+  @Test(
+          groups = "mercury_htmlTitleSet",
+          dataProviderClass = MercurySEODataProvider.class,
+          dataProvider = "htmlTitles"
+  )
+  public void mercury_htmlTitleSet(Page page, String expectedTitle) {
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/testcases/mercurytests/HTMLTitleTests.java]: MAJOR Rename this method name to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS00100)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Comment by tomasznapieralski** [comment]
LGTM

**Comment by vforge** [comment]
LGTM
